src/librbd/ManagedLock.h

```diff
@@ -0,0 +1,181 @@
+
+namespace librbd {
+
+class Watcher;
+
+template <typename ImageCtxT>
+class ManagedLock {
+public:
+  static const std::string WATCHER_LOCK_TAG;
+
+  static ManagedLock *create(librados::IoCtx& ioctx, ContextWQ *work_queue,
+                             const std::string& oid, Watcher *watcher) {
+    return new ManagedLock(ioctx, work_queue, oid, watcher);
+  }
+
+  ManagedLock(librados::IoCtx& ioctx, ContextWQ *work_queue,
+              const std::string& oid, Watcher *watcher);
+  virtual ~ManagedLock();
+
+  bool is_lock_owner() const;
+
+  void shut_down(Context *on_shutdown);
+  void acquire_lock(Context *on_acquired, bool try_lock = false);
+  void release_lock(Context *on_released);
+  void reacquire_lock(Context *on_reacquired = nullptr);
+
+  virtual void shutdown_handler(int r, Context *on_finish);
+  virtual void pre_acquire_lock_handler(Context *on_finish);
+  virtual void post_acquire_lock_handler(int r, Context *on_finish);
+  virtual void pre_release_lock_handler(bool shutting_down,
+                                        Context *on_finish);
+  virtual void post_release_lock_handler(bool shutting_down, int r,
+                                          Context *on_finish);
+
+  void assert_locked(librados::ObjectWriteOperation *op, ClsLockType type);
+
+  bool is_shutdown() const {
+    Mutex::Locker l(m_lock);
+    return is_shutdown_locked();
+  }
+
+  bool is_locked_state() const {
+    return m_state == STATE_LOCKED;
+  }
+
+  static bool decode_lock_cookie(const std::string &tag, uint64_t *handle);
+
+protected:
+
+  /**
+   * @verbatim
+   *
+   *       <start>
+   *          |
+   *          |
+   *          v           (acquire_lock)
+   *       UNLOCKED -----------------------------------------> ACQUIRING
+   *          ^                                                    |
+   *          |                                                    |
+   *      RELEASING                                                |
+   *          |                                                    |
+   *          |                                                    |
+   *          |                    (release_lock)                  v
+   *    PRE_RELEASING <----------------------------------------- LOCKED
+   *
+   * <LOCKED state>
+   *    |
+   *    v
+   * REACQUIRING -------------------------------------> <finish>
+   *    .                                                 ^
+   *    .                                                 |
+   *    . . . > <RELEASE action> ---> <ACQUIRE action> ---/
+   *
+   * <UNLOCKED/LOCKED states>
+   *    |
+   *    |
+   *    v
+   * PRE_SHUTTING_DOWN ---> SHUTTING_DOWN ---> SHUTDOWN ---> <finish>
+   *
+   * @endverbatim
+   */
+  enum State {
+    STATE_UNLOCKED,
+    STATE_LOCKED,
+    STATE_ACQUIRING,
+    STATE_WAITING_FOR_REGISTER,
+    STATE_REACQUIRING,
+    STATE_RELEASING,
+    STATE_PRE_SHUTTING_DOWN,
+    STATE_SHUTTING_DOWN,
+    STATE_SHUTDOWN,
+
+    // ExclusiveLock Specific
```

Comment:
I think all of these states could be made generic enough so that you don't need to call out that these are exclusive lock states oddly embedded in the base class. Derived classes can provided a ```initialization_handler``` virtual method, ```post_acquiring``` and ```pre_releasing```, and rename waiting for peer to ```waiting_for_lock``` (i.e. rbd-mirror might want to "wait for the lock" by starting a timer to re-attempt the acquire).